### PR TITLE
Handle comments and overhaul `Styles.Defs`

### PR DESCRIPTION
**Pull request by GregMefford**
I think this is just about ready to go - all the tests are passing and things are working how I'd expect, so feel free to go through and nitpick the implementation. 😁 I did not try to compare it too much against the spike you did yet, but I did run it against our internal codebase and it seems to be working well.

There are probably just a few odds and ends to update the docs to no longer say that Defs is disabled, bump the version, etc. which I can knock out tomorrow morning.

### CODE REVIEW

File: lib/style.ex [hunk 1]
```diff
@@ -30,4 +30,84 @@ defmodule Styler.Style do
...
+
+  then this function would manipulate the `comments` such that they would end
+  up being formatted like this:
+
+      # This is arg1
+      # This is arg2
+      def(arg1, arg2), do: :ok
+
+  Obviously, it would then be up to the developer to rewrite the comments to
+  that they're more descriptive about what they are referring to.
```
Comment (novaugust) [inline lib/style.ex:60]:
```suggestion
  Set the line of all comments with `line` in `range_start..range_end` to instead have line `range_start`
```
I think you wrote this comment with the idea that it'd take ast ;)

File: lib/style.ex [hunk 2]
```diff
@@ -30,4 +30,84 @@ defmodule Styler.Style do
...
+  then this function would manipulate the `comments` such that they would end
+  up being formatted like this:
+
+      # Positive numbers are good
+      def(arg1, arg2) when is_integer(arg1) and arg1 >= 0, do: :ok
+
+      # Negative numbers are are bad
+      def(arg1, arg2), do: :error
+
+      # This comment comes at the end
```
Comment (novaugust) [inline lib/style.ex:102]:
```suggestion
  Change the `line` of all comments with `line` in `range` by adding `delta` to it
```

File: lib/style/defs.ex [hunk 3]
```diff
@@ -41,60 +41,107 @@ defmodule Styler.Style.Defs do
...
+    {_fn_name, head_meta, _children} = head
+    first_line = meta[:line]
+    last_line = head_meta[:closing][:line]
+
+    comments =
+      if first_line == last_line do
+        ctx.comments
+      else
+        Style.displace_comments(ctx.comments, first_line..last_line)
+      end
```
Comment (novaugust) [inline lib/style/defs.ex:61]: i think you can use this branch to just skip all work, not just moving coments!
first clause is `{:skip, zipper, ctx)`, second line is all the displace/flatten/replace work

File: lib/style/defs.ex [hunk 4]
```diff
@@ -41,60 +41,107 @@ defmodule Styler.Style.Defs do
...
+        #
+        [{{:__block__, do_meta, [:do]}, {_, body_meta, _}}] = body
+        def_line = def_meta[:line]
+        do_line = do_meta[:line]
+        end_line = body_meta[:closing][:line] || do_meta[:line]
+        {:keyword, def_line, do_line, end_line}
+      end
 
-    if def_start_line == def_end_line do
+    if def_line == end_line do
```
Comment (novaugust) [inline lib/style/defs.ex:94]: nit:
hmm, could cut down on `if block else` in here by getting just the end line above, and leaving the `do_line` determination inside of the `if block`
like:
```elixir
def_line = def_meta[:line]
end_line =
  if line = def_meta[:do][:line] || def_meta[:end_of_expression][:line] do
    line
  else
    [{{:__block__, do_meta, [:do]}, {_, body_meta, _}}] = body
    body_meta[:closing][:line] || do_meta[:line]
  end

cond do
  def_line == end_line -> :skip
  do_line = def_meta[:do][:line] ->
    # we're in a def do end
    ... 
    {:skip, ...}
  true ->
    # we're in a keyword
    ...
    {:skip, ...}
end
```
sorry, had to write that out to see how it compares. shrug. i think it is fine to not do the `{node, comments} = if ..` just for the sake of only having one skip statement.

File: lib/style/defs.ex [hunk 5]
```diff
@@ -41,60 +41,107 @@ defmodule Styler.Style.Defs do
...
+          head = flatten_head(head, def_line)
+          body = update_all_meta(body, collapse_lines(set_to_def_line))
+          comments = Style.displace_comments(ctx.comments, def_line..end_line)
 
-      body = update_all_meta(body, body_meta_rewriter)
+          {{def, def_meta, [head, body]}, comments}
+        end
 
-      # There won't be any defs deeper in here, so lets skip ahead if we can
       # @TODO this skips checking the body, which can be incorrect if therey's a `quote do def do ...` inside of it
```
Comment (novaugust) [inline lib/style/defs.ex:128]: lol this would be easily fixed if we just changed it from `skip` to `next`. 
¯\\_(ツ)_/¯
be smart to add tests for it tho. i can do it in a followup

File: test/style/defs_test.exs [hunk 6]
```diff
@@ -11,17 +11,203 @@
 defmodule Styler.Style.DefsTest do
   use Styler.StyleCase, style: Styler.Style.Defs, async: true
 
+  describe "elixir ast dependencies we rely on" do
```
Comment (novaugust) [inline test/style/defs_test.exs:14]: let's remove this whole describe block

**Review by novaugust — APPROVED** [review]
:shipit: so great. NOW we're cooking with controversial grease baby!

would you update the moduledoc on defs? pretty sure it still says it isn't used ;)

likewise readme needs an update, maybe a new section for non-credo styles we do?

File: lib/style/defs.ex (hunk 4 above)
Comment (novaugust) [reply lib/style/defs.ex:94]: >i think it is fine to not do the {node, comments} = if .. just for the sake of only having one skip statement.

this sentence hard to read. let me try again.

i think only having one return site, and so matching {node, comments} out of an if statement, is maybe harder to follow than just having each branch of the statement have its own return value

Comment (GregMefford) [reply lib/style/defs.ex:94]: Yeah I tried it both ways and landed where I did because I thought there was more duplication between the branches than there actually turned out to be, because both branches are basically calling the same functions but with slightly different parameters.

I'll give it a try like you're sketching out above and see how it looks.

Comment (GregMefford) [reply lib/style/defs.ex]: Ok yeah I think I do like it better with a `cond` so let's go with that. ❤️

**Comment by GregMefford** [comment]
Got the docs and versions bumped, LMK if that stuff all looks good to go. We can probably also merge your PR and release them together in the same version though.

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,16 +1,23 @@
 # Changelog
 
+## v0.3.0
```
Comment (novaugust) [inline CHANGELOG.md:3]:
```suggestion
## main
```
or some other thing so people dont' come to our page and think a release is out?
unless you're kicking it as soon as you merge =)

(i've just been pushing release bumps straight to main when i'm ready to do it tee hee)

**Comment by novaugust** [comment]
everything looks good to me, but wanna add something to the readme telling people about this style? or do we just roll with the "go look at moduledocs" section and call it good?
right now everything's kinda explicitly documented via the `credo rules we replace` table, so this'll be the first thing that _isnt_ if we choose to leave it out

File: CHANGELOG.md (hunk 7 above)
Comment (GregMefford) [reply CHANGELOG.md:3]: Yeah I've also seen `v0.3.0-dev` in the version so it's clear that it's kind of a collection point for things. I guess either way, you have to take that step to push a commit when you do the final tagging. I'm cool with this pattern though, so we just always write our own changelog for each unreleased feature, then we cut a release separately.

File: lib/style/defs.ex (hunk 5 above)
Comment (novaugust) [reply lib/style/defs.ex]: yeah i don't really have a lot of worries about macros that write code we should style not getting styled, because as we all know: the first rule of macros is, don't write macros

**Comment by GregMefford** [comment]
Yeah I chose not to include it there because that section is about superseded Credo rules and there is no Credo rule for this.

**Comment by novaugust** [comment]
well then :shipit:
